util/design/lib/OtpMemImg.py

```diff
@@ -0,0 +1,384 @@
+
+        item_size = self.config['partitions'][part_idx]['items'][item_idx][
+            'size']
+        random_or_hexvalue(item, 'value', item_size * 8)
+        self.config['partitions'][part_idx]['items'][item_idx]['value'] = item[
+            'value']
+
+        log.info('> Adding item {} with size {}B and value:'.format(
+            item['name'], item_size))
+        fmt_str = '{:0' + str(item_size * 2) + 'x}'
+        value_str = fmt_str.format(item['value'])
+        bytes_per_line = 8
+        j = 0
+        while value_str:
+            # Print out max 64bit per line
+            line_str = ''
+            for k in range(bytes_per_line):
+                num_chars = min(len(value_str), 2)
+                line_str += value_str[-num_chars:]
+                if k < bytes_per_line - 1:
+                    line_str += ' '
+                value_str = value_str[:len(value_str) - num_chars]
+            log.info('  {:06x}: '.format(j) + line_str)
+            j += bytes_per_line
+
+    def override_data(self, img_config):
+        '''Override specific partition items'''
+
+        if 'partitions' not in img_config:
+            raise RuntimeError('Missing partitions key in configuration.')
+
+        for part in img_config['partitions']:
+            self.merge_part_data(part)
+            log.info('Overriding values of {} partition.'.format(part['name']))
+            for item in part['items']:
+                self.merge_item_data(part, item)
+
+    def streamout_partition(self, part_idx):
+        '''Scramble and stream out partition data as a list of bytes'''
+
+        part_name = self.config['partitions'][part_idx]['name']
+        log.info('Streamout of partition {}'.format(part_name))
+
+        part_size = self.config['partitions'][part_idx]['size']
+        part_offset = self.config['partitions'][part_idx]['offset']
+        # First chop up all items into individual bytes.
+        data_bytes = [0] * part_size
+        # Annotation is propagated into the hexfile as comments
+        annotation = ['unallocated'] * part_size
+        # Need to keep track of defined items for the scrambling.
+        # Undefined regions are left blank (0x0) in the memory.
+        defined = [False] * part_size
+        for item in self.config['partitions'][part_idx]['items']:
+            for k in range(item['size']):
+                idx = item['offset'] - part_offset + k
+                annotation[idx] = part_name + ': ' + item['name']
+                if 'value' in item:
+                    data_bytes[idx] = ((item['value'] >> (8 * k)) & 0xFF)
+                    defined[idx] = True
+
+        # Reshape this into 64bit blocks (this must be aligned at this point)
+        if len(data_bytes) % 8:
+            raise RuntimeError('Partitions must be 64bit aligned')
+
+        data_blocks = []
+        data_block_defined = []
+        for k, b in enumerate(data_bytes):
```

Comment:
If `part_size` isn't a multiple of 8, I think this loses some data. Do we need an assertion or check somewhere?

Reply:
part size is guaranteed to be a multiple of 8, since this is checked in the `OtpMemImg` constructor.
I'll add an assertion.